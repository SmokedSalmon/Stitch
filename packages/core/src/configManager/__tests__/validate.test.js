import getValidate from '../validate'

describe('validate', () => {
  test('type object', () => {
    const schema = {
      $id: 'mockedSchema_type_object',
      type: 'object',
      properties: {
        expectedKey: {
          type: 'object',
          properties: {}
        },
        requiredKey: {
          type: 'object',
          properties: {}
        }
      },
      required: ['requiredKey']
    }
    const validate = getValidate(schema)

    expect(validate({ requiredKey: {} })).toBe(true)

    try {
      validate('nonObject')
    } catch (error) {
      expect(error).toMatchSnapshot('failed with non-object')
    }

    try {
      validate({ unexpectedKey: {} })
    } catch (error) {
      expect(error).toMatchSnapshot('failed with unexpected key')
    }

    try {
      validate({ expectedKey: {} })
    } catch (error) {
      expect(error).toMatchSnapshot('failed with missing required key')
    }
  })

  test('type object failed with wrong schema', () => {
    const validate = getValidate({
      $id: 'mockedSchema_type_object_failed',
      type: 'object'
    })

    try {
      validate({})
    } catch (error) {
      expect(error).toMatchSnapshot('missing "properties" define')
    }
  })

  test('type array', () => {
    const defaultValue = [{ defaultItem: {} }]
    const config = {}
    const schema = {
      $id: 'mockedSchema_type_array',
      type: 'object',
      properties: {
        arrayProperty: {
          type: 'array',
          items: {
            type: 'object',
            properties: {}
          },
          default: defaultValue
        }
      }
    }
    const validate = getValidate(schema)

    expect(validate(config)).toBe(true)
    // `default` schema will side effects the validated config object
    expect(config.arrayProperty).toEqual(defaultValue)

    try {
      validate({
        arrayProperty: 'nonArray'
      })
    } catch (error) {
      expect(error).toMatchSnapshot('failed with non-array')
    }
  })

  test('type array with minItems', () => {
    const schema = {
      $id: 'mockedSchema_type_array_minItems',
      type: 'array',
      items: {
        type: 'object',
        properties: {}
      },
      minItems: 2
    }
    const validate = getValidate(schema)

    expect(validate([{}, {}])).toBe(true)

    try {
      validate([{}])
    } catch (error) {
      expect(error).toMatchSnapshot('failed')
    }
  })

  test('type array with uniqueItems', () => {
    const schema = {
      $id: 'mockedSchema_type_array_uniqueItems',
      type: 'array',
      items: { type: 'string' },
      uniqueItems: true
    }
    const validate = getValidate(schema)

    expect(validate(['a', 'b'])).toBe(true)

    try {
      validate(['a', 'a'])
    } catch (error) {
      expect(error).toMatchSnapshot('failed')
    }
  })

  test('type array failed with wrong schema', () => {
    try {
      getValidate({
        $id: 'mockedSchema_type_array_wrong1',
        type: 'array',
        items: {},
        default: 'nonArray'
      })([])
    } catch (error) {
      expect(error).toMatchSnapshot('wrong "default" define')
    }

    try {
      getValidate({
        $id: 'mockedSchema_type_array_wrong2',
        type: 'array'
      })([])
    } catch (error) {
      expect(error).toMatchSnapshot('missing "items" define')
    }

    try {
      getValidate({
        $id: 'mockedSchema_type_array_wrong3',
        type: 'array',
        items: {},
        minItems: 'nonNumber'
      })([])
    } catch (error) {
      expect(error).toMatchSnapshot('wrong "minItems" define')
    }

    try {
      getValidate({
        $id: 'mockedSchema_type_array_wrong4',
        type: 'array',
        items: {},
        uniqueItems: 'nonBoolean'
      })([])
    } catch (error) {
      expect(error).toMatchSnapshot('wrong "uniqueItems" define')
    }
  })

  test('type string', () => {
    const defaultValue = 'defaultString'
    const config = {}
    const schema = {
      $id: 'mockedSchema_type_string',
      type: 'object',
      properties: {
        stringProperty: {
          type: 'string',
          default: defaultValue
        }
      }
    }
    const validate = getValidate(schema)

    expect(validate(config)).toBe(true)
    // `default` schema will side effects the validated config object
    expect(config.stringProperty).toBe(defaultValue)

    try {
      getValidate({
        $id: 'mockedSchema_type_string_failed',
        type: 'string'
      })({ nonString: true })
    } catch (error) {
      expect(error).toMatchSnapshot('failed with non-string')
    }
  })

  test('type string with enum', () => {
    const schema = {
      $id: 'mockedSchema_type_string_enum',
      type: 'string',
      enum: [
        'expectedEnum1',
        'expectedEnum2'
      ]
    }
    const validate = getValidate(schema)

    expect(validate('expectedEnum1')).toBe(true)

    try {
      expect(validate('unexpectedEnum'))
    } catch (error) {
      expect(error).toMatchSnapshot('failed with unexpected enum')
    }
  })

  test('type string with pattern', () => {
    const schema = {
      $id: 'mockedSchema_type_string_pattern',
      type: 'string',
      pattern: '^expectedPattern$'
    }
    const validate = getValidate(schema)

    expect(validate('expectedPattern')).toBe(true)

    try {
      expect(validate('unexpectedPattern'))
    } catch (error) {
      expect(error).toMatchSnapshot('failed with unexpected pattern')
    }
  })

  test('type string failed with wrong schema', () => {
    try {
      getValidate({
        $id: 'mockedSchema_type_string_wrong1',
        type: 'string',
        default: ['nonString']
      })('string')
    } catch (error) {
      expect(error).toMatchSnapshot('wrong "default" define')
    }

    try {
      getValidate({
        $id: 'mockedSchema_type_string_wrong2',
        type: 'string',
        enum: 'nonArray'
      })('string')
    } catch (error) {
      expect(error).toMatchSnapshot('wrong "enum" define')
    }

    try {
      getValidate({
        $id: 'mockedSchema_type_string_wrong3',
        type: 'string',
        pattern: '['
      })('string')
    } catch (error) {
      expect(error).toMatchSnapshot('wrong "pattern" define')
    }
  })

  test('type number', () => {
    const schema = {
      $id: 'mockedSchema_type_array_wrong4',
      type: 'number',
      maximum: 200
    }
    const validate = getValidate(schema)

    expect(validate(123)).toBe(true)

    try {
      validate('nonNumber')
    } catch (error) {
      expect(error).toMatchSnapshot('failed with non-number')
    }

    try {
      validate(999)
    } catch (error) {
      expect(error).toMatchSnapshot('failed with maximum')
    }
  })

  test('type number failed with wrong schema', () => {
    try {
      getValidate({
        $id: 'mockedSchema_type_number_wrong1',
        type: 'number',
        maximum: 'nonNumber'
      })(123)
    } catch (error) {
      expect(error).toMatchSnapshot('wrong "maximum" define')
    }
  })

  test('type boolean', () => {
    const defaultValue = true
    const config = {}
    const schema = {
      $id: 'mockedSchema_type_boolean',
      type: 'object',
      properties: {
        booleanProperty: {
          type: 'boolean',
          default: defaultValue
        }
      }
    }
    const validate = getValidate(schema)

    expect(validate(config)).toBe(true)
    // `default` schema will side effects the validated config object
    expect(config.booleanProperty).toBe(defaultValue)

    try {
      validate(
        getValidate({
          $id: 'mockedSchema_type_boolean_failed',
          type: 'boolean'
        })('nonBoolean'))
    } catch (error) {
      expect(error).toMatchSnapshot('failed with non-boolean')
    }
  })

  test('type boolean failed with wrong schema', () => {
    try {
      getValidate({
        $id: 'mockedSchema_type_boolean_wrong1',
        type: 'boolean',
        default: 'nonBoolean'
      })(true)
    } catch (error) {
      expect(error).toMatchSnapshot('wrong "default" define')
    }
  })

  test('type undefined', () => {
    const schema = {
      $id: 'mockedSchema_type_undefined',
      type: 'undefined'
    }
    const validate = getValidate(schema)

    expect(validate(undefined)).toBe(true)

    try {
      validate('nonUndefined')
    } catch (error) {
      expect(error).toMatchSnapshot('failed with non-undefined')
    }
  })

  test('type unsupported', () => {
    try {
      getValidate({
        $id: 'mockedSchema_type_unsupported',
        type: 'unsupported'
      })(() => {})
    } catch (error) {
      expect(error).toMatchSnapshot()
    }
  })

  test('missing type define', () => {
    try {
      getValidate({
        $id: 'mockedSchema_type_unsupported'
      })(() => {})
    } catch (error) {
      expect(error).toMatchSnapshot()
    }
  })

  test('oneOf', () => {
    const schema = {
      $id: 'mockedSchema_oneOf',
      type: 'string',
      oneOf: [
        { type: 'string' },
        { type: 'number' }
      ]
    }
    const validate = getValidate(schema)

    expect(validate('expectedString')).toBe(true)
    expect(validate(333)).toBe(true)

    try {
      expect(validate({}))
    } catch (error) {
      expect(error).toMatchSnapshot('failed with unmatched oneOf')
    }
  })

  test('oneOf failed with wrong schema', () => {
    try {
      getValidate({
        $id: 'mockedSchema_oneOf_wrong1',
        type: 'string',
        oneOf: 'nonArray'
      })('string')
    } catch (error) {
      expect(error).toMatchSnapshot('wrong "oneOf" define')
    }
  })

  test('format hostname', () => {
    const schema = {
      $id: 'mockedSchema_format_hostname',
      type: 'string',
      format: 'hostname'
    }
    const validate = getValidate(schema)

    expect(validate('mocked-host.name.com')).toBe(true)

    try {
      expect(validate('unsupported hostname'))
    } catch (error) {
      expect(error).toMatchSnapshot('failed')
    }
  })

  test('format ipv4', () => {
    const schema = {
      $id: 'mockedSchema_format_ipv4',
      type: 'string',
      format: 'ipv4'
    }
    const validate = getValidate(schema)

    expect(validate('127.0.0.1')).toBe(true)

    try {
      expect(validate('unsupported ipv4'))
    } catch (error) {
      expect(error).toMatchSnapshot('failed')
    }
  })

  test('format ipv6', () => {
    const schema = {
      $id: 'mockedSchema_format_ipv6',
      type: 'string',
      format: 'ipv6'
    }
    const validate = getValidate(schema)

    expect(validate('::1')).toBe(true)

    try {
      expect(validate('unsupported ipv6'))
    } catch (error) {
      expect(error).toMatchSnapshot('failed')
    }
  })

  test('format uri', () => {
    const schema = {
      $id: 'mockedSchema_format_uri',
      type: 'string',
      format: 'uri'
    }
    const validate = getValidate(schema)

    expect(validate('mocked-protocol://mocked/path')).toBe(true)

    try {
      expect(validate('unsupported uri'))
    } catch (error) {
      expect(error).toMatchSnapshot('failed')
    }
  })

  test('format unsupported', () => {
    try {
      getValidate({
        $id: 'mockedSchema_format_unsupported',
        type: 'string',
        format: 'unsupported'
      })(() => {})
    } catch (error) {
      expect(error).toMatchSnapshot()
    }
  })
})
