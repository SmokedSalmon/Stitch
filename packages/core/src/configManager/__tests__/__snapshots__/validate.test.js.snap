// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate format hostname: failed 1`] = `[TypeError: Config validate error: config should be the format of hostname of schema node mockedSchema_format_hostname]`;

exports[`validate format ipv4: failed 1`] = `[TypeError: Config validate error: config should be the format of ipv4 of schema node mockedSchema_format_ipv4]`;

exports[`validate format ipv6: failed 1`] = `[TypeError: Config validate error: config should be the format of ipv6 of schema node mockedSchema_format_ipv6]`;

exports[`validate format unsupported 1`] = `[TypeError: Schema error: unsupported format of mockedSchema_format_unsupported]`;

exports[`validate format uri: failed 1`] = `[TypeError: Config validate error: config should be the format of uri of schema node mockedSchema_format_uri]`;

exports[`validate missing type define 1`] = `[TypeError: Schema error: missing type define of mockedSchema_type_unsupported]`;

exports[`validate oneOf failed with wrong schema: wrong "oneOf" define 1`] = `[TypeError: Schema error: oneOf rule of mockedSchema_oneOf_wrong1 should be an array]`;

exports[`validate oneOf: failed with unmatched oneOf 1`] = `[TypeError: Config validate error: config should match one of schema mockedSchema_oneOf/oneOf]`;

exports[`validate type array failed with wrong schema: missing "items" define 1`] = `[TypeError: Schema error: missing items define of type "array" schema node mockedSchema_type_array_wrong2]`;

exports[`validate type array failed with wrong schema: wrong "default" define 1`] = `[TypeError: Schema error: default define of type "array" schema node mockedSchema_type_array_wrong1 should be an Array]`;

exports[`validate type array failed with wrong schema: wrong "minItems" define 1`] = `[TypeError: Schema error: minItems define of type "array" schema node mockedSchema_type_array_wrong3 should be a number]`;

exports[`validate type array failed with wrong schema: wrong "uniqueItems" define 1`] = `[TypeError: Schema error: uniqueItems define of type "array" schema node mockedSchema_type_array_wrong4 should be a boolean]`;

exports[`validate type array with minItems: failed 1`] = `[TypeError: Config validate error: config should contains at least 2 item]`;

exports[`validate type array with uniqueItems: failed 1`] = `[TypeError: Config validate error: items of config should be unique]`;

exports[`validate type array: failed with non-array 1`] = `[TypeError: Config validate error: config.arrayProperty should be an array]`;

exports[`validate type boolean failed with wrong schema: wrong "default" define 1`] = `[TypeError: Schema error: default define of type "boolean" schema node mockedSchema_type_boolean_wrong1 should be a boolean]`;

exports[`validate type boolean: failed with non-boolean 1`] = `[TypeError: Config validate error: config should be a boolean]`;

exports[`validate type number failed with wrong schema: wrong "maximum" define 1`] = `[TypeError: Schema error: maximum define of type "number" schema node mockedSchema_type_number_wrong1 should be a number]`;

exports[`validate type number: failed with maximum 1`] = `[TypeError: Config validate error: the maximum value of config is 200]`;

exports[`validate type number: failed with non-number 1`] = `[TypeError: Config validate error: config should be a number]`;

exports[`validate type object failed with wrong schema: missing "properties" define 1`] = `[TypeError: Schema error: missing properties define of type "object" schema node mockedSchema_type_object_failed]`;

exports[`validate type object: failed with missing required key 1`] = `[TypeError: Config validate error: missing required property "requiredKey" of config]`;

exports[`validate type object: failed with non-object 1`] = `[TypeError: Config validate error: config should be an object]`;

exports[`validate type object: failed with unexpected key 1`] = `[TypeError: Schema error: unexpected key or property at mockedSchema_type_object/unexpectedKey]`;

exports[`validate type string failed with wrong schema: wrong "default" define 1`] = `[TypeError: Schema error: default define of type "string" schema node mockedSchema_type_string_wrong1 should be a string]`;

exports[`validate type string failed with wrong schema: wrong "enum" define 1`] = `[TypeError: Schema error: enum define of type "string" schema node mockedSchema_type_string_wrong2 should be an Array]`;

exports[`validate type string failed with wrong schema: wrong "pattern" define 1`] = `[TypeError: Schema error: pattern define of type "string" mockedSchema_type_string_wrong3 should be a RegExp constructable string]`;

exports[`validate type string with enum: failed with unexpected enum 1`] = `[TypeError: Config validate error: config should be a value in the enum mockedSchema_type_string_enum/enum]`;

exports[`validate type string with pattern: failed with unexpected pattern 1`] = `[TypeError: Config validate error: config should match the pattern of mockedSchema_type_string_pattern/pattern]`;

exports[`validate type string: failed with non-string 1`] = `[TypeError: Config validate error: config should be a string]`;

exports[`validate type undefined: failed with non-undefined 1`] = `[TypeError: Config validate error: config accept undefined, but received string]`;

exports[`validate type unsupported 1`] = `[TypeError: Schema error: unsupported type of mockedSchema_type_unsupported]`;
